thread.c

```diff
@@ -545,9 +545,125 @@ static void thread_libevent_process(int fd, short which, void *arg) {
     }
 }
 
-/* Which thread we assigned a connection to most recently. */
+/* The thread we assigned a connection to most recently. */
 static int last_thread = -1;
 
+/* This function simply returns next_thread based on round-robin */
+static LIBEVENT_THREAD *select_next_thread(void)
+{
+    int tid = (last_thread + 1) % settings.num_threads;
+    last_thread = tid;
+    return threads + tid;
+}
+
+/* This function selects a thread to process requests from a given connection.
+ * Thread selection is either based on the (existing) default mechanism,
+ * round-robin based thread selection, or based on NAPI_ID. NAPI_ID based
+ * selection creates an association between a (hardware) queue on a NIC and an
+ * application thread. This mapping streamlines the flow of data from the NIC
+ * to an application thread by having an application thread source data from
+ * the same hardware queue which serves up traffic for many connections
+ */
+static void select_thread(int sfd, LIBEVENT_THREAD **thread_selected)
+{
+    LIBEVENT_THREAD *thread, *selected_thread = NULL;
+    socklen_t len;
+    int napi_id;
+
+    /* if napi-id based thread selection is not enabled, use
+     * default round-robin based thread selection
+     */
+	if (!settings.napi_id_based_thread_selection) {
+        selected_thread = select_next_thread();
+        goto out;
+    }
+
+    /* Now proceed with NAPI-ID based thread selection */
+
+    /* first check if there is pre-existing NAPI_ID => Thread
+     * association, if so - use that thread, otherwise select unassigned
+     * or unassociated thread 
+     */
+#ifndef SO_INCOMING_NAPI_ID
+#define SO_INCOMING_NAPI_ID 56
+#endif /* SO_INCOMING_NAPI_ID */
+    len = sizeof(napi_id);
+    if (getsockopt(sfd, SOL_SOCKET, SO_INCOMING_NAPI_ID,
+		   &napi_id, &len) == -1) {
+        perror("getsockopt SO_INCOMING_NAPI_ID");
+        /* Disable napi_id  based thread selection */
+        settings.napi_id_based_thread_selection = false;
```

Comment:
is there no way to detect this earlier? can it fail permanently during runtime?

Reply:
This will fails only if you arr running on older linux kernel 4.12 since SO_INCOMING_NAPI_ID socket option was added in 4.12 upstream linux kernel.

Once it fails, it will always fail during runtime, hence disabling napi_id based thread selection mechanism forever.

To detect it earlier, at least we have to make one call on accepted connection, hence it was done here and if fails - disable the napi_id based selection logic.

Please let me know if I am not answering to your question or if I misunderstood your comment

Reply:
Can this test be done on the listening socket during startup?

Reply:
Yes. So you want to perform this test on listening socket and if fails, disable the feature completely and don;t do it for every accepted.child socket.

Sure. will make that change.

Reply:
I think it should try to bail out the entire application or loudly print at least. Unless there's a reason not to, a user trying to enable an option that then doesn't work should be communicated as early as possible. If we can check during the listen stage that should be early enough to bail out.